devcheck

```diff
@@ -15,9 +16,29 @@
 
 # This script runs all tests and examples and verifies they work as expected.
 
-[[ "${BASH_SOURCE[0]}" != "${0}" ]] && EXIT=return || EXIT=exit
+python3 -m unittest discover --start-directory . --pattern '*_test.py' --verbose
+
+if [ $? -ne 0 ]; then
+  echo "Test failed with code $?"
+  exit
+fi
 
-python3 -m unittest discover -s . -p '*_test.py' -v || { RESULT=$? ; echo -e "\n\n Test failed with code $RESULT\n\n" ; $EXIT $RESULT ; }
+tests_passed=true
 
- eval $(find examples/ -name 'run.sh' -printf 'echo -e "\\n\\n+++++++++++ Running: %p" && %p && ') echo -e '+++++\\n\\nChecks: OK' || { RESULT=$? ; echo -e "+++++\n\nChecks: ERROR (status: $RESULT) above" ; $EXIT $RESULT ; }
+for test_script in `find examples/ -type f -name run.sh`; do
+  test_directory="$( dirname "$test_script" )"
+  example_name="$( basename "$test_directory" )"
+  echo "Running $example_name"
+  if $test_script; then
+    echo "Checks: OK"
+  else
+    echo "Checks: ERROR (status: $?)"
```

Comment:
The script should exit with `$?` so that we can easily use it in further scripts. I suggest we do `tests_passed=$?` and then `$EXIT $tests_passed`

(Also applies to the unit tests above.)

Reply:
> The script should exit with $?

Which `$?`   
Do you want it to stop right away?

If so, when you check CircleCI, you'll have no idea whether the other tests passed or not.  
e.g. if the first test fails, you wonder "did something break everything or is only that broken?"

Reply:
tl;dr: I think a sensible policy would be to run as many tests/checks as we can, and to exit with the exit code of the first failing test, or 0 if they all pass. (Alternatively, just exit with code 1 if any tests fail, rather than worrying about preserving the exact failing exit code.)

My thinking:
* I want the script to only exit with a 0 code if all the tests/checks pass. Otherwise, some non-zero code would be good. That's the most important thing.
* I agree that running as many tests as possible in one sweep would be great to save developer time.